noman.el

```diff
@@ -69,141 +70,115 @@
-                                                    (point-min)
-                                                    (point-max)))
-                        (indent-code-rigidly (point-min) (point-max) 4)
-                        (buffer-string))))
-                   (versioninfo-p (not (string= (string-trim versioninfo) ""))))
-          (goto-char (point-max))
-          (insert "\nIMPLEMENTATION\n")
-          (insert versioninfo))))
+       ((string-suffix-p " is a shell builtin" type)
+        (call-process shell-file-name nil t nil "-c" (format "'%s help -m'" cmd)))
+       ((string-suffix-p " not found" type)
+        (user-error "Command '%s' not found" prefix))
+       (t (unless (= (apply #'call-process prefix nil t nil
+                            `(,@(cdr tokens) "--help"))
+                     0)
+            (erase-buffer)
+            (call-process cmd nil t nil "help")
+            (replace-regexp-in-region "." "" (point-min) (point-max)))
+          (when-let ((versioninfo
+                      (save-excursion
+                        (with-temp-buffer
+                          (unless (= (call-process prefix nil t nil  "--version") 0)
+                            (erase-buffer)
+                            (call-process prefix nil t nil "version")
+                            (replace-regexp-in-region "." "" (point-min) (point-max)))
+                          (indent-code-rigidly (point-min) (point-max) 4)
+                          (buffer-string))))
+                     (versioninfo-p (not (string-empty-p (string-trim versioninfo)))))
+            (goto-char (point-max))
+            (insert "\nIMPLEMENTATION\n")
+            (insert versioninfo))))
       (ansi-color-apply-on-region (point-min) (point-max))
-      (read-only-mode t)
-      (noman-mode)
-      (setq noman--buttons (noman--make-buttons buffer cmd))
-      (setq noman--last-command cmd)
       (goto-char (point-min))
-      (display-buffer buffer))))
+      (unless (derived-mode-p 'noman-mode) (noman-mode))
+      (setq noman--buttons (noman--buttonize cmd)
+            noman--last-command cmd
+            header-line-format (string-join (split-string cmd " ")  " > "))
+      (display-buffer (current-buffer) '(display-buffer-reuse-window)))))
+
+(defun noman--read-cmd ()
+  "Return executable command from user prompt."
+  (cl-loop for path in (cl-loop for p in (exec-path)
+                                append (if (file-directory-p p)
+                                           (directory-files p t "^[^.].*")
+                                         (list p)))
+           when (and  (file-executable-p path))
+           collect (file-name-nondirectory path) into commands
+           finally return (completing-read "Program: " commands nil t)))
+
+;;;###autoload
+(defun noman (cmd)
+  "Return to parsed command line help for the command CMD."
+  (interactive (list (noman--read-cmd)))
+  (noman--buffer cmd)
+  (push cmd noman--history))
```

Comment:
Now the main command only pushes the cmd to history if it succeeds.
Most of the logic is delegated to internal functions.